autogpt_platform/backend/backend/data/execution.py

```diff
@@ -10,62 +20,144 @@
-from backend.data.queue import AsyncRedisEventBus, RedisEventBus
 from backend.server.v2.store.exceptions import DatabaseError
-from backend.util import mock, type
+from backend.util import mock
+from backend.util import type as type_utils
 from backend.util.settings import Config
 
+from .block import BlockData, BlockInput, CompletedBlockOutput
+from .db import BaseDbModel
+from .includes import EXECUTION_RESULT_INCLUDE, GRAPH_EXECUTION_INCLUDE
+from .model import GraphExecutionStats, NodeExecutionStats
+from .queue import AsyncRedisEventBus, RedisEventBus
 
-class GraphExecutionEntry(BaseModel):
-    user_id: str
-    graph_exec_id: str
-    graph_id: str
-    graph_version: int
-    start_node_execs: list["NodeExecutionEntry"]
+T = TypeVar("T")
 
+config = Config()
 
-class NodeExecutionEntry(BaseModel):
+
+# -------------------------- Models -------------------------- #
+
+
+ExecutionStatus = AgentExecutionStatus
+
+
+class GraphExecutionMeta(BaseDbModel):
     user_id: str
-    graph_exec_id: str
+    started_at: datetime
+    ended_at: datetime
+    cost: Optional[int] = Field(..., description="Execution cost in credits")
+    duration: float = Field(..., description="Seconds from start to end of run")
+    total_run_time: float = Field(..., description="Seconds of node runtime")
+    status: ExecutionStatus
     graph_id: str
-    node_exec_id: str
-    node_id: str
-    block_id: str
-    data: BlockInput
+    graph_version: int
+    preset_id: Optional[str] = None
 
+    @staticmethod
+    def from_db(_graph_exec: AgentGraphExecution):
+        now = datetime.now(timezone.utc)
+        start_time = _graph_exec.startedAt or _graph_exec.createdAt
+        end_time = _graph_exec.updatedAt or now
+        duration = (end_time - start_time).total_seconds()
+        total_run_time = duration
+
+        try:
+            stats = type_utils.convert(_graph_exec.stats or {}, dict[str, Any])
+        except ValueError:
+            stats = {}
+
+        duration = stats.get("walltime", duration)
+        total_run_time = stats.get("nodes_walltime", total_run_time)
+
+        return GraphExecutionMeta(
+            id=_graph_exec.id,
+            user_id=_graph_exec.userId,
+            started_at=start_time,
+            ended_at=end_time,
+            cost=stats.get("cost", None),
+            duration=duration,
+            total_run_time=total_run_time,
+            status=ExecutionStatus(_graph_exec.executionStatus),
+            graph_id=_graph_exec.agentGraphId,
+            graph_version=_graph_exec.agentGraphVersion,
+            preset_id=_graph_exec.agentPresetId,
+        )
 
-ExecutionStatus = AgentExecutionStatus
 
-T = TypeVar("T")
+class GraphExecution(GraphExecutionMeta):
+    inputs: dict[str, Any]
+    outputs: dict[str, list[Any]]
+    node_executions: list["NodeExecutionResult"]
 
+    @staticmethod
+    def from_db(_graph_exec: AgentGraphExecution):
+        from .block import BlockType, get_block
```

Comment:
Won't this cause the repetitive cached import call? is this causing circular import ?

Reply:
I'll merge it with the module-level import from `.block`